src/_library.ts

```diff
@@ -1,28 +1,54 @@
-export type DynamicLibraryType =
-  | "bool"
-  | "i8"
-  | "u8"
-  | "i16"
-  | "u16"
-  | "i32"
-  | "u32"
-  | "i65"
-  | "u64"
-  | "f32"
-  | "f64"
-  | "pointer"
-  | { struct: Readonly<Array<DynamicLibraryType>> };
+import { PlatformPointer } from "./_types.ts";
+type DynamicLibraryStructType = { readonly struct: Readonly<Array<DynamicLibraryType>> };
+type DynamicLibraryTypeMap = {
+  "bool": boolean;
+  "i8": number;
+  "u8": number;
+  "i16": number;
+  "u16": number;
+  "i32": number;
+  "u32": number;
+  "i64": number | bigint;
+  "u64": number | bigint;
+  "f32": number;
+  "f64": number;
+  "pointer": PlatformPointer<unknown>;
+  "void": void;
+};
+export type DynamicLibraryType = Exclude<keyof DynamicLibraryTypeMap, "void"> | DynamicLibraryStructType;
+export type DynamicLibraryResultType = keyof DynamicLibraryTypeMap | DynamicLibraryStructType;
 
 export type DynamicLibraryFunctionInterface = {
   readonly parameters: readonly DynamicLibraryType[];
-  readonly result: DynamicLibraryType | "void";
+  readonly result: DynamicLibraryResultType;
 };
 
 export type DynamicLibraryInterface = Record<string, DynamicLibraryFunctionInterface>;
 
-// TODO: Can we improve the type inference here for symbols?
-export type DynamicLibrary<T> = {
-  symbols: Record<keyof T, (...args: unknown[]) => unknown>;
+type ToNativeType<T extends DynamicLibraryType = DynamicLibraryType> = T extends DynamicLibraryStructType ? BufferSource
+  : DynamicLibraryTypeMap[Exclude<T, DynamicLibraryStructType>];
+type StaticForeignSymbolReturnType<T extends DynamicLibraryFunctionInterface> = T["result"] extends
+  DynamicLibraryStructType ? Uint8Array
+  : DynamicLibraryTypeMap[Exclude<T["result"], DynamicLibraryStructType>];
+type ToNativeParameterTypes<T extends readonly DynamicLibraryType[]> =
+  //
+  [(T[number])[]] extends [T] ? ToNativeType<T[number]>[]
+    : [readonly (T[number])[]] extends [T] ? readonly ToNativeType<T[number]>[]
+    : T extends readonly [...DynamicLibraryType[]] ? {
+        [K in keyof T]: ToNativeType<T[K]>;
+      }
+    : never;
+type FromForeignFunction<T extends DynamicLibraryFunctionInterface> = T["parameters"] extends readonly []
+  ? () => StaticForeignSymbolReturnType<T>
+  : (
+    ...args: ToNativeParameterTypes<T["parameters"]>
+  ) => StaticForeignSymbolReturnType<T>;
+type StaticForeignLibraryInterface<T extends DynamicLibraryInterface> = {
+  [K in keyof T]: FromForeignFunction<T[K]>;
+};
+
+export type DynamicLibrary<T extends DynamicLibraryInterface> = {
+  symbols: StaticForeignLibraryInterface<T>;
```

Comment:
Too much of this feels copied from the `deno` source code and I don't really see the great benefit here. The FFI layer isn't visible to the end user and the code that calls the FFI is normally completely generated. In the case of the `Quit` functions that was a copy and paste error and from the looks of it your change to the `generator.ts` to make the `symbols` a pure `const` export seems to actually fix the problem with calling functions by the wrong name.

Reply:
These code help me to detect whether we use an pointer as buffer, as changes made at `src/SDL_ttf/functions.ts`.

Reply:
`src/SDL_ttf/functions.ts` is a completely generated file though so I don't see the major win here. Even if the type is correct it doesn't mean invoking the actual type marshalling will work.

Reply:
A procedure to convert pointer at struct to arraybuffer is introduced in `src/SDL_ttf/functions.ts`, because type checking indicated that pointer from `_data` struct class may be used as input parameter of struct type, which expects arraybuffer instead.

Reply:
@smack0007 If i revert this change, will you suggest me to remove `denoToPlatformStruct` or other feature?

Reply:
@jingkaimori No I want the rest of the changes. The `denoToPlatformStruct` is a definite improvement.

Reply:
How can we indicate that `_data` field should be transferred to arraybuffer without type checking?

Reply:
That's what line 1038 in `tools/codegen/generators.ts` does.

Reply:
If we can rely on some 'magic' code generator rather than type system, why not use `any` type for `_library` variable?

Reply:
It basically is `any`. `export type DynamicLibrary<T> = { symbols: Record<keyof T, (...args: unknown[]) => unknown>; }`. 

TBH I'm just about done here. If you want me to accept the MR than take this part out otherwise we seem to have a fundamental difference in opinion. I'm happy to accept the rest but I don't want this part.